### PR TITLE
Display and support editing the area outside the map bounds

### PR DESCRIPTION
**Pull request by pchote**
Fixes one of the few limitations that the new editor has vs the old.  Also fixes the missing tiles that used to appear at the right/bottom edges of the screen in TS (I was sure we had a ticket on this, but can't find it).

Depends on #8337.

### CODE REVIEW

File: OpenRA.Mods.Common/EditorBrushes/EditorActorBrush.cs [hunk 1]
```diff
@@ -27,6 +27,7 @@ namespace OpenRA.Mods.Common.Widgets
 	public class EditorActorBrush : IEditorBrush
 	{
 		public readonly ActorInfo Actor;
+		public readonly CVec[] footprint;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/EditorBrushes/EditorActorBrush.cs]: > [SA1307] Public and internal fields must start with an upper-case letter.

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

**Comment by pchote** [comment]
Fixed and rebased.

**Comment by LipkeGu** [comment]
Crashes when placing or removing actors which are outside from the map (bottom) ?!?!?

```Tiberian Sun Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.1.7601 Service Pack 1)
Runtime Version: .NET CLR 4.0.30319.18444
Exception of type `System.IndexOutOfRangeException`: Der Index war außerhalb des Arraybereichs.
   bei OpenRA.CellLayer`1.get_Item(CPos cell) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Map\CellLayer.cs:Zeile 82.
   bei OpenRA.Mods.Common.Widgets.EditorDefaultBrush.HandleMouseInput(MouseInput mi)
   bei OpenRA.Mods.Common.Widgets.EditorViewportControllerWidget.HandleMouseInput(MouseInput mi)
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Widgets\Widget.cs:Zeile 354.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Widgets\Widget.cs:Zeile 345.
   bei OpenRA.Widgets.Ui.HandleInput(MouseInput mi) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Widgets\Widget.cs:Zeile 90.
   bei OpenRA.DefaultInputHandler.<>c__DisplayClass7.<OnMouseInput>b__6() in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Input\InputHandler.cs:Zeile 49.
   bei OpenRA.Sync.CheckSyncUnchanged[T](World world, Func`1 fn) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Sync.cs:Zeile 182.
   bei OpenRA.DefaultInputHandler.OnMouseInput(MouseInput input) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Input\InputHandler.cs:Zeile 49.
   bei OpenRA.Renderer.Sdl2.Sdl2Input.PumpInput(IInputHandler inputHandler) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Renderer.Sdl2\Sdl2Input.cs:Zeile 86.
   bei OpenRA.Renderer.Sdl2.Sdl2GraphicsDevice.PumpInput(IInputHandler inputHandler) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Renderer.Sdl2\Sdl2GraphicsDevice.cs:Zeile 334.
   bei OpenRA.Graphics.Renderer.EndFrame(IInputHandler inputHandler) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Graphics\Renderer.cs:Zeile 156.
   bei OpenRA.Game.RenderTick() in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Game.cs:Zeile 517.
   bei OpenRA.Game.Loop() in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Game.cs:Zeile 618.
   bei OpenRA.Game.Run() in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Game.cs:Zeile 636.
   bei OpenRA.Program.Run(String[] args) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Support\Program.cs:Zeile 111.
   bei OpenRA.Program.Main(String[] args) in d:\OpenRA-editor-cordon\OpenRA-editor-cordon\OpenRA.Game\Support\Program.cs:Zeile 39.
```

**Comment by pchote** [comment]
Fixed the crash, and added support for resources outside the map bounds (which I forgot earlier).

> When placing tiles outside the bounds it will produce black tiles on the placed tiles

I'm not sure what you mean by this and can't reproduce it.  The editor doesn't let you place tiles outside the map bounds?

**Comment by LipkeGu** [comment]
these black tiles on the map

**Comment by pchote** [comment]
Good catch: that was a bug in the underlying `Tileshape.Diamond` code.  Added a commit to fix it here.

**Comment by penev92** [comment]
Well code looks good. Will need some testing though :+0.5:

**Comment by penev92** [comment]
I get
>System.IndexOutOfRangeException was unhandled
  Message=Index was outside the bounds of the array.
  StackTrace:
       at OpenRA.CellLayer`1.get_Item(CPos cell) in \OpenRA.Game\Map\CellLayer.cs:line 82
       at OpenRA.Mods.Common.Traits.EditorResourceLayer.ResourceDensityAt(CPos c)
       at OpenRA.Mods.Common.Traits.EditorResourceLayer.UpdateDirtyTile(CPos c)
       at OpenRA.Mods.Common.Traits.EditorResourceLayer.Render(WorldRenderer wr)
       at OpenRA.Graphics.TerrainRenderer.Draw(WorldRenderer wr, Viewport viewport) in \OpenRA.Game\Graphics\TerrainRenderer.cs:line 44

When trying to place resources on one of the edge-most tiles on the top/bottom edge of a map in both TS and D2k.

Also 

Top edge looks OK (crashes all the same though).

**Comment by pchote** [comment]
Fixed and rebased.

**Comment by penev92** [comment]
Looks OK to me now :+1:

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/32c0443416b0f91fa92319851f87cd538b174dcc...5f8a5a855fee148d9e3d635377c296baab5b891c)

**Pull request merged**